cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/SourceControlManagementHttpHandler.java

```diff
@@ -163,6 +171,38 @@ public void pushApp(FullHttpRequest request, HttpResponder responder,
     }
   }
 
+  /**
+   * Pull the requested application from linked repository and deploy in current namespace.
+   */
+  @POST
+  @Path("/apps/{app-id}/pull")
+  public void pullApp(FullHttpRequest request, HttpResponder responder,
+                      @PathParam("namespace-id") String namespaceId,
+                      @PathParam("app-id") final String appId) throws Exception {
+    checkSourceControlFeatureFlag();
+    ApplicationReference appRef = validateAppReference(namespaceId, appId);
+
+    try {
+      ApplicationRecord appRecord = sourceControlService.pullAndDeploy(appRef);
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(appRecord));
+    } catch (NoChangesToPullException | InvalidArtifactException e) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, e.getMessage());
+    } catch (ArtifactNotFoundException e) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, e.getMessage());
+    } catch (ConflictException e) {
+      responder.sendString(HttpResponseStatus.CONFLICT, e.getMessage());
+    } catch (UnauthorizedException e) {
+      responder.sendString(HttpResponseStatus.FORBIDDEN, e.getMessage());
+    } catch (IOException e) {
+      LOG.error("Error reading request body for creating app {}.", appId);
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, String.format(
+        "Error while reading json request body for app %s.", appId));
```

Comment:
this doesn't look like the right error message, there is no request body. Do we need to differentiate between this and the general Exception case?

Reply:
Removed this one since there's actually no request body.